lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -166,6 +173,118 @@ static bool isTypeTooComplex(Type t) {
+}
+
+/// Checks if a second substitution list is a expanded version of
+/// the first substitution list.
+/// This is the case if at least one of the substitution type in Subs2 is
+/// "bigger" than the corresponding substitution type in Subs1.
+/// Type T2 is "smaller" than type T1 if T2 is structurally contained in T1.
+static bool growingSubstitutions(SubstitutionList Subs1,
+                                 SubstitutionList Subs2) {
+  assert(Subs1.size() == Subs2.size());
+  // Perform component-wise substitution comparisions.
+  for (unsigned idx = 0, e = Subs1.size(); idx < e; ++idx) {
+    auto Type1 = Subs1[idx].getReplacement()->getCanonicalType();
+    auto Type2 = Subs2[idx].getReplacement()->getCanonicalType();
+    // Replacement types should be concrete.
+    assert(!Type1->hasArchetype());
+    assert(!Type2->hasArchetype());
+    if (Type2 == Type1)
+      continue;
+    if (isContainedIn(Type2, Type1))
+      continue;
+    if (isContainedIn(Type1, Type2)) {
+      DEBUG(llvm::dbgs() << "Type:\n"; Type1.dump();
+            llvm::dbgs() << "is contained in type:\n"; Type2.dump());
+      DEBUG(llvm::dbgs() << "SubstitutionList[" << idx
+                         << "] has got bigger since last time.\n");
+      return true;
+    }
+    // None of the types is contained in the other type.
+    // They are not comparable in this sense.
+  }
+
+  // The substitition list is not growing.
+  return false;
+}
+
+/// Checks whether specializing a given generic apply would create an infinite
+/// cycle in the generic specializations graph. This can be the case if there is
+/// a loop in the specialization graph and generic parameters at each iteration
+/// of such a loop are getting bigger and bigger.
+/// The specialization graph is represented by means of SpecializationInformation.
+/// We use this meta-information about specializations to detect cycles in this
+/// graph.
+static bool createsInfiniteSpecializationLoop(ApplySite Apply) {
+  if (!Apply)
+    return false;
+  auto *Callee = Apply.getCalleeFunction();
+  SILFunction *Caller = nullptr;
+  Caller = Apply.getFunction();
+
+  // Name of the function to be specialized.
+  auto GenericFunc = Callee;
+
+  DEBUG(llvm::dbgs() << "\n\n\nChecking for a specialization cycle:\n"
+                     << "Caller: " << Caller->getName() << "\n"
+                     << "Callee: " << Callee->getName() << "\n";
+        llvm::dbgs() << "Substitutions:\n";
+        for (auto Sub: Apply.getSubstitutions()) {
+          Sub.getReplacement()->dump();
+        });
+
+  auto *CurSpecializationInfo = Apply.getSpecializationInfo();
+  if (CurSpecializationInfo) {
+    DEBUG(llvm::dbgs() << "Scan call-site's history\n");
+  } else if (Caller->isSpecialization()) {
+    CurSpecializationInfo = Caller->getSpecializationInfo();
+    DEBUG(llvm::dbgs() << "Scan caller's specialization history\n");
+  }
+
+  while (CurSpecializationInfo) {
```

Comment:
I think the checks are fast enough here. 

Let's say we have a specialization chain of depth N. Checking the components if the current callee occurs on this chain is very cheap. And in situations it is not the case and so we are done. If the callee occurs on the specialization chain, then it is potentially a specialization loop and we need to perform more expensive checks to see if substitutions grow. But IMHO it happens only in rare cases and thus it is very unlikely to see a real compilation performance hit here.